### PR TITLE
feat(query-bar): add sign in modal with sign in flow to query bar ai input

### PR DESCRIPTION
**Pull request by gribnoysup**
This is the last part extracted from the POC branch, this patch adds an sign in modal and related logic to the query bar that handles user sign in:

https://github.com/mongodb-js/compass/assets/5036933/00c94e1f-fdaf-4620-a353-81af7f2f13ee

This needs to be run against the backend with the auth changes applied (https://github.com/10gen/compass-mongodb-com/pull/72)

### CODE REVIEW

File: packages/compass-query-bar/src/stores/atlas-signin-reducer.ts [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  return async (dispatch, getState, { atlasService }) => {
+    if (getState().atlasSignIn.state === 'success') {
+      return;
+    }
+    dispatch(cancelSignIn());
+    const { id, signal } = getAbortSignal();
+    dispatch({ type: AtlasSignInActions.RestoringStart, id });
+    dispatch({
+      type: (await atlasService.isAuthenticated({ signal }).catch(() => {
+        // For the initial state check if failed to check auth for any reason we
```
Comment (gribnoysup) [inline packages/compass-query-bar/src/stores/atlas-signin-reducer.ts:143]: This should check if we caught a cancellation error and not dispatch in that case

File: packages/compass-query-bar/src/stores/atlas-signin-reducer.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+      } catch (err) {
+        if (!signal.aborted) {
+          // Only show error toast if sign in wasn't aborted by the user
+          openToast('atlas-sign-in-error', {
+            variant: 'important',
+            title: 'Sign in failed',
+            description: (err as Error).message,
+          });
+        }
+        dispatch({
```
Comment (gribnoysup) [inline packages/compass-query-bar/src/stores/atlas-signin-reducer.ts:190]: Also needs to be skipped on cancel

File: packages/atlas-service/src/main.ts [hunk 3]
```diff
@@ -124,12 +124,18 @@ export class AtlasService {
     ]);
   }
 
-  static async isAuthenticated(): Promise<boolean> {
+  static async isAuthenticated({
```
Comment (mcasimir) [inline packages/atlas-service/src/main.ts:127]: We can add a test also here?

File: packages/compass-query-bar/src/components/generative-ai/ai-sign-in-modal.tsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            Atlas users can now quickly generate queries and aggregations with
+            MongoDB’s AI-powered features, available today in Compass.
+          </p>
+          <p>
+            This feature is an experiment and may give inaccurate responses. You
+            can help make it better by leaving feedback.
+          </p>
+          <p>
+            To understand how your data is used with AI partners,{' '}
+            <Link href="https://example.com">learn more in the docs</Link>.
```
Comment (mcasimir) [inline packages/compass-query-bar/src/components/generative-ai/ai-sign-in-modal.tsx:86]: To start here we could link the home of the Compass documentation and have a ticket to update it later.

File: packages/compass-query-bar/src/stores/ai-query-reducer.ts [hunk 5]
```diff
@@ -39,6 +41,8 @@ export const enum AIQueryActionTypes {
   ShowInput = 'compass-query-bar/ai-query/ShowInput',
   HideInput = 'compass-query-bar/ai-query/HideInput',
   ChangeAIPromptText = 'compass-query-bar/ai-query/ChangeAIPromptText',
+  ShowOptIn = 'compass-query-bar/ai-query/ShowOptIn',
```
Comment (mcasimir) [inline packages/compass-query-bar/src/stores/ai-query-reducer.ts:44]: ShowLogin / HideLogin?

File: packages/compass-query-bar/src/stores/ai-query-reducer.ts [hunk 6]
```diff
@@ -39,6 +41,8 @@ export const enum AIQueryActionTypes {
...
+  HideOptIn = 'compass-query-bar/ai-query/HideOptIn',
```
Comment (gribnoysup) [inline packages/compass-query-bar/src/stores/ai-query-reducer.ts:45]:
```suggestion
```

**Review by mcasimir — APPROVED** [review]

File: packages/compass-query-bar/src/stores/ai-query-reducer.ts (hunk 6 above)
Comment (mcasimir) [reply packages/compass-query-bar/src/stores/ai-query-reducer.ts:45]: Ah ok  :)

Comment (gribnoysup) [reply packages/compass-query-bar/src/stores/ai-query-reducer.ts:45]: Yeah, moved them to sign in reducer, and forgot to clean up here, good catch!

File: packages/atlas-service/src/main.ts (hunk 3 above)
Comment (gribnoysup) [reply packages/atlas-service/src/main.ts:129]: Added, thanks for highlighting

File: packages/compass-query-bar/src/components/generative-ai/ai-sign-in-modal.tsx (hunk 4 above)
Comment (gribnoysup) [reply packages/compass-query-bar/src/components/generative-ai/ai-sign-in-modal.tsx]: figma designs actually had a link that I didn't notice, added, thanks for catching

**Review by Anemy — COMMENTED** [review]
Looking good, couple small questions and things that caught my eye. I dig the unmounting changes. I'll give it a try now.

File: packages/compass-collection/src/components/collection/collection.tsx [hunk 7]
```diff
@@ -166,7 +166,6 @@ const Collection: React.FunctionComponent<CollectionProps> = ({
           views={views}
           activeTabIndex={activeSubTab}
           onTabClicked={(tabIdx) => onSubTabClicked(tabIdx, tabs[tabIdx])}
-          mountAllViews
```
[the remark this answers is not in the document]
Comment (Anemy) [reply packages/compass-collection/src/components/collection/collection.tsx:169]: Ah found one, looks like after changing text in the validation tab, then changing tabs, and back to the validation tab, the changes are erased. 
We're probably re-fetching on every mount and overriding the state there.

**Review by addaleax — APPROVED** [review]
Didn't try it out but the source looks good, I like @anemy's suggestions but I guess they could also be addressed separately if you want to get this merged as soon as possible?

**Comment by gribnoysup** [comment]
Ah, yes, just saw those, will check before merging

File: packages/compass-collection/src/components/collection/collection.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (gribnoysup) [reply packages/compass-collection/src/components/collection/collection.tsx:169]: This is the same how it works right now, we can open a bug ticket, but that's not related to these changes

File: packages/compass-query-bar/src/stores/atlas-signin-reducer.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+      if ((await atlasService.isAuthenticated({ signal })) === false) {
+        await atlasService.signIn({ signal });
+      }
+      const user = await atlasService.getUserInfo({ signal });
+      openToast('atlas-sign-in-success', {
+        variant: 'success',
+        title: `Signed in as ${user.login}`,
+        timeout: 10_000,
+      });
+      dispatch({ type: AtlasSignInActions.Success });
```
[the remark this answers is not in the document]
Comment (gribnoysup) [reply packages/compass-query-bar/src/stores/atlas-signin-reducer.ts:215]: This is what `getInitialSignInState` action does in the createStore, but we'd need a more robust mechanism than that for the case of switching existing tabs, yes. I'll keep it as is in this patch: the worst thing that will happen if the state is out of sync right now is that you will get a modal multiple times, you won't need to sign in again, just click the button. And I'm already planning a slight refactor in the follow-up that should make managing this easier, so will handle this case there too.

File: packages/compass-collection/src/components/collection/collection.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (gribnoysup) [reply packages/compass-collection/src/components/collection/collection.tsx:169]: Done in https://github.com/mongodb-js/compass/pull/4661/commits/ecf1c1ec863c5f3311cd29153c96e93a12d4abd7

File: packages/compass-query-bar/src/components/generative-ai/ai-sign-in-modal.tsx [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        <Subtitle className={titleStyles}>
+          Build faster with AI&nbsp;<Badge>Experimental</Badge>
+        </Subtitle>
+        <div className={descriptionStyles}>
+          <p>
+            Atlas users can now quickly generate queries and aggregations with
+            MongoDB’s AI-powered features, available today in Compass.
+          </p>
+          <p>
+            This feature is an experiment and may give inaccurate responses. You
```
[the remark this answers is not in the document]
Comment (gribnoysup) [reply packages/compass-query-bar/src/components/generative-ai/ai-sign-in-modal.tsx]: Done in https://github.com/mongodb-js/compass/pull/4661/commits/38cb77ecc934daeb5ccf8f5a064b0f94d7dc4a96

File: packages/compass-query-bar/src/components/generative-ai/ai-sign-in-modal.tsx [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            <Link href="https://example.com">learn more in the docs</Link>.
+          </p>
+        </div>
+        <div className={controlsStyles}>
+          <Button
+            variant="primary"
+            onClick={onSignInClick}
+            className={buttonStyles}
+            disabled={isSignInInProgress}
+            // TODO: will have to update leafygreen for that
```
[the remark this answers is not in the document]
Comment (gribnoysup) [reply packages/compass-query-bar/src/components/generative-ai/ai-sign-in-modal.tsx:98]: Done in https://github.com/mongodb-js/compass/pull/4661/commits/38cb77ecc934daeb5ccf8f5a064b0f94d7dc4a96

**Pull request merged**